### ISSUE TITLE
corrupted on verification

### ISSUE THREAD

**Issue by dblmask** [nicoboss/nsz#52]
I'm using master brach 5b3afac and I have **hash misssmatch** on some files after being compressed and verified with the following command:

`python nsz.py -C -l 22 -V -w -o "dir" "file.nsp"
`

On some files I get:

> Solid compressing (level 22) 
> [VERIFY NSZ] [010085700abc8800][v65536].nsz
> [EXISTS]     758d1bc445e70398d5c44fa7abf4a1a7.cnmt.xml
> [CORRUPTED]  233c4b75c9b4192394b3d833cfa704de.ncz
> nut exception: Verification detected hash missmatch
> 
> [VERIFY NSZ] [0100EA80032EA000][v0].nsz
> [CORRUPTED]  dc10021f535d07b3daf50908c462d7d5.nca
> nut exception: Verification detected hash missmatch!
> 

And I have to comment line 46 on nsz/\_\_init\_\_.py: raise
because the program hangs waiting for something and never ends, I had to press Ctrl+C which is bad on scripts.

Why is happenning the hash missmatch and how to fix this?

**Comment by nicoboss** [comment]
“CORRUPTED: Verification detected hash mismatch!” means the compressed game is corrupted and can't be decompressed back into the original game. Things like this is exactly why always using -V is highly recommended especially on development builds as otherwise you will just end up with completely useless corrupted games if something went wrong. The verification option decompresses and sha256 hashes every game after compression and compares the resulting hash with the hash inside the CNMT file also used by real hardware to verify your games. If decompression doesn't result in the exact bit identical NCA file it detects a hash mismatch and raises an unhandled exception because this is a very critical situation.

To compress without getting interrupted by don't use -V while compressing and later verify them all using -V only where it will perform the exact same hash check but continues on errors as when manually requesting a verification it's assumed you read the output while by just adding the -V parameter must people won't even care what's written to the console except there’s an exception. I will probably add a command line option specifying if exceptions on hash mismatch should be thrown in the near future and put them inside the exception summary at the very end as normal exception. But it's quite questionable if the batch process should really continue after such a huge exception.

Now you might wonder what probably cause this to happen after NSZ latest commit being somewhat stable for such a long time:
The latest commit https://github.com/nicoboss/nsz/commit/5b3afac2efeb6221ffa8411a12165229c21b4589 on master fixes a huge issue related to XCI compression. This required changes at the very core design of booth compression and decompression. This major change is currently barely tested as I have to study for my university which starts next Monday. I'm still awaiting testing feedback from @popy2k14. For more information what was changed please take a look at https://github.com/nicoboss/nsz/issues/49

It would be highly appreciated if you could let me know which games break. Please also try if verifying this corrupted game compressed with latest commit (just choose one you know will be corrupt and compress it without the -V option) verified on v3.1.1 (using only the -V option to verify compressed games) will be flag flagged as CORRUPTED too. If yes, the bug is in compression otherwise in decompression. Providing such information would make debugging and fixing this bug much faster and easier.

In the meantime, for you as end user the easy solution for your problem is obviously to just use the latest stable release v3.1.1 or https://github.com/nicoboss/nsz/commit/372024b08bf097ba70b7eb55e89a27e3f5767c40 as the latest commit only affects XCI/XCZ games and there is no reason to use it for NSP games.

Even though I’m directly working on master I will usually try to avoid pushing unstable poorly tested commits to master but this time it was very important as it prevented most of the XCI games from being compressed properly. Generally, I always recommend using latest release instead of latest commit for actual use except there are very good reasons because releases are always way more stable and much better tested.

**Comment by popy2k14** [comment]
@nicoboss sorry for giving you too much hope.
New year eve is over and i started work again, so sadly not  much freetime to test it soon.

**Comment by dblmask** [comment]
Thanks for that complete answer, I wish I could help you coding. I have tried compressing with version 2.1.1 , 3.1.1 and 5b3afac
First without verification, then with verification.
Some games that get corruption with 2.1.1 get no corruption with 3.1.1.

Here's a sample:
```
./nsz.py -C -l 22 -w -o "dir/" "file.nsp"

Solid compressing (level 22) [010085700abc8800][v65536].nsp -> [010085700abc8800][v65536].nsz
[ADDING]     758d1bc445e70398d5c44fa7abf4a1a7.cnmt.xml 1678 bytes to NSP
[ADDING]     233c4b75c9b4192394b3d833cfa704de.ncz 62884352 bytes to NSP
Compressed 70.79559951575871% 62884352 -> 44519354  - 233c4b75c9b4192394b3d833cfa704de.nca
[ADDING]     0936e5dd76ec1477d0f7b59b3b5c8a58.nca 198656 bytes to NSP
[ADDING]     d21af68bccaa0551277201fc6ab1cf59.nca 81920 bytes to NSP
[ADDING]     a9e94e3f2262d77e1743f253c05388be.nca 143360 bytes to NSP
[ADDING]     758d1bc445e70398d5c44fa7abf4a1a7.cnmt.nca 4096 bytes to NSP
```

```
./nsz.py -C -l 22 -V -w -o "dir/" "file.nsp"

Solid compressing (level 22) [010085700abc8800][v65536].nsp -> [010085700abc8800][v65536].nsz
[ADDING]     758d1bc445e70398d5c44fa7abf4a1a7.cnmt.xml 1678 bytes to NSP
[ADDING]     233c4b75c9b4192394b3d833cfa704de.ncz 62884352 bytes to NSP
Compressed 70.79559951575871% 62884352 -> 44519354  - 233c4b75c9b4192394b3d833cfa704de.nca
[ADDING]     0936e5dd76ec1477d0f7b59b3b5c8a58.nca 198656 bytes to NSP
[ADDING]     d21af68bccaa0551277201fc6ab1cf59.nca 81920 bytes to NSP
[ADDING]     a9e94e3f2262d77e1743f253c05388be.nca 143360 bytes to NSP
[ADDING]     758d1bc445e70398d5c44fa7abf4a1a7.cnmt.nca 4096 bytes to NSP
[VERIFY NSZ] [010085700abc8800][v65536].nsz
[EXISTS]     758d1bc445e70398d5c44fa7abf4a1a7.cnmt.xml
[CORRUPTED]  233c4b75c9b4192394b3d833cfa704de.ncz
nut exception: Verification detected hash missmatch
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "~/nsz-3.1.1/nsz/__init__.py", line 41, in solidCompressTask
    verify(outFile, True, [statusReport, id], pleaseNoPrint)
  File "~/nsz-3.1.1/nsz/__init__.py", line 67, in verify
    NszVerify(filePath, raiseVerificationException, statusReportInfo, pleaseNoPrint)
  File "~/nsz-3.1.1/nsz/NszDecompressor.py", line 43, in verify
    __decompressNsz(filePath, None, False, raiseVerificationException, statusReportInfo, pleaseNoPrint)
  File "~/nsz-3.1.1/nsz/NszDecompressor.py", line 184, in __decompressNsz
    __decompressContainer(container, None, fileHashes, write, raiseVerificationException, statusReportInfo, pleaseNoPrint)
  File "~/nsz-3.1.1/nsz/NszDecompressor.py", line 87, in __decompressContainer
    raise Exception("Verification detected hash missmatch")
Exception: Verification detected hash missmatch
^C
```
This ones are still giving me hash missmatch.

- [010055d009f78000][v0].nsp
- [010055D009F78800][v196608].nsp
- [0100633007D48000][v0].nsp
- [010002C00C270000][v0].nsp
- [010085700abc8800][v65536].nsp
- [010003B00D3A2000][v0].nsp
- [01005DE00D05C000][v0].nsp
- [010061D00DB74000][v0].nsp
- [0100BF500207C000][v0].nsp
- [0100EA80032EA000][v0].nsp


I hope this helps bugfixing

**Comment by husky73** [comment]
As far as I can tell, some NSPs contain already corrupt NCAs (their hash does not match that contained in CNMT). Consequently, NSZ also fails the verification, being the NSZ verification based on the hashes contained in CNMT. I suggest checking NSP before compressing it.

**Comment by deadair3210** [comment]
Only improperly dumped or modified NSP files have corrupt NCA. If the stock files had the issue from the start the Switch would never accept them as it would constantly think that they are corrupted

**Comment by eXhumer** [comment]
0100633007D48000 works fine.

**Comment by eXhumer** [comment]
010055D009F78000 compresses fine and both the input and output are verified with `-V`.

**Comment by eXhumer** [comment]
Are we sure this isn't user error or users using NCA unverified files? Because I just compressed 2 of the files which apparently corrupts on compression and both the inputs and outputs are verified.

**Comment by eXhumer** [comment]
```
Solid compressing (level 18) C:\Users\eXhumer\Documents\GitProjects\nsz\[010055D009F78000][v0].nsp -> C:\Users\eXhumer\Documents\GitProjects\nsz\[010055D009F78000][v0].nsz
[ADDING]     010055d009f780000000000000000008.cert 1792 bytes to NSP
[ADDING]     010055d009f780000000000000000008.tik 704 bytes to NSP
[ADDING]     8e079aa8f8ef2e83e5230db02b65b488.ncz 11789664256 bytes to NSP
Compressed 88.61219693074185% 11789664256 -> 10447080508  - 8e079aa8f8ef2e83e5230db02b65b488.nca
[ADDING]     67962d3737ca185be9a406ed39f26e41.nca 204800 bytes to NSP
[ADDING]     2300492db9f7ab07da4835969c9e17d3.cnmt.nca 3584 bytes to NSP
[ADDING]     a938d9d513d6dab11fe61a893eb2bc09.nca 1192448 bytes to NSP
[VERIFY NSZ] C:\Users\eXhumer\Documents\GitProjects\nsz\[010055D009F78000][v0].nsz
[EXISTS]     010055d009f780000000000000000008.cert
[EXISTS]     010055d009f780000000000000000008.tik
[VERIFIED]   8e079aa8f8ef2e83e5230db02b65b488.ncz
[VERIFIED]   67962d3737ca185be9a406ed39f26e41.nca
[EXISTS]     2300492db9f7ab07da4835969c9e17d3.cnmt.nca
[VERIFIED]   a938d9d513d6dab11fe61a893eb2bc09.nca
Done!
```

**Comment by dblmask** [comment]
> Are we sure this isn't user error or users using NCA unverified files as inputs? Because I just compressed 2 of the files which apparently corrupts on compression and both the inputs and outputs are verified.

Which version are you using?
How do you verify NCA ?

**Comment by eXhumer** [comment]
Latest commit
Run the script with just -V command (`python3 nsz.py -V file.nsp`) and see if the input file is verified or not.

**Comment by eXhumer** [comment]
If possible, please provide steps you did to re-produce the errors, as I was unable to get any error regarding verification.

**Comment by nicoboss** [comment]
This issue seems to be fixed with latest NSZ v4.0 please reopen if you can still recreate it.
